### DISCUSSION TITLE
Thread model for embedding WAMR in a multi-threaded environment

### DISCUSSION THREAD

**Discussion by adazh** [bytecodealliance/wasm-micro-runtime#3697] [Q&A] [answered]
Hi! We are trying to use WAMR in a multi-threaded environment. From what I understand so far, we can:
- Create a WAMR engine (via `wasm_runtime_init`) once in one process.
- Load the `.wasm` module (via `wasm_runtime_load`) once and share the `wasm_module_t` object between threads. It's also thread-safe to access the module object concurrently. Does this model work when loading multiple modules?
- What I noticed is that we'll have to call `wasm_runtime_init_thread_env` when initializing the module instance via `wasm_runtime_instantiate` from a different thread than where the module is loaded. Otherwise, the `wasm_runtime_lookup_function` will fail to find functions in the module.

Could anyone possibly confirm whether the understanding/observations above are correct?

I'm also wondering whether it's feasible to share a module instance (`wasm_module_inst_t`) between threads? By holding on to different `exec_env`s for example.

Thanks! Appreciate any thoughts/pointers!

**Comment by wenyongh** [comment, marked as answer]
Hi, I try to answer the questions:
- yes, had better call wasm_runtime_init once
- accessing wasm_module_t in multiple threads when multi-module feature is enabled: normally it is good as long as the module is loaded, note that no lock is acquired when accessing module's multi-module related fields, e.g., `module->import_module_list`.
- no need to call wasm_runtime_init_thread_env if the thread is created by runtime or runtime API, e.g. using wasm_runtime_spawn_thread. Only call it if the thread is created by host native itself (e.g. pthread_create), and please call it at the beginning of thread callback, and call wasm_runtime_destroy_thread_env at the end of thread callback.
- had better not share one wasm_module_inst_t in different threads, not sure whether there will be unexpected behavior. You may try accessing it with different exec_envs and in un-overlapped timeslots, but I am not so sure whether it is safe - at least we shouldn't call wasm functions simultaneously in different threads.

**Reply by adazh** [reply]
Thanks for explaining. This is helpful!
